### DISCUSSION TITLE
Why `--userns=keep-id` doesn't work?

### DISCUSSION THREAD

**Discussion by abitrolly** [containers/podman#16258] [Q&A]
```
✗ mkdir xxx
✗ podman run -it --rm --security-opt label=disable --userns=keep-id -v $PWD/xxx:/xxx gitpod
gitpod ~ $ ls -la /xxx
total 8
drwxr-xr-x.  2 anatoli anatoli 4096 Oct 22 04:32 .
dr-xr-xr-x. 20 root    root    4096 Oct 22 04:32 ..
gitpod ~ $ echo $UID
33333
gitpod ~ $ touch /xxx/aaa
touch: cannot touch '/xxx/aaa': Permission denied
```
Why it doesn't map to my $UID? These are values from my host.
```
✗ ps faux | grep "b\ash"
133332     75825  0.1  0.1  12392  9188 pts/0    Ss+  07:32   0:00      \_ bash
✗ echo $UID
1000
```

```
 podman version
Client:       Podman Engine
Version:      4.2.1
API Version:  4.2.1
Go Version:   go1.18.5
Built:        Wed Sep  7 22:58:19 2022
OS/Arch:      linux/amd64
```

**Comment by mheon** [comment]
THe ownership of `.` from the `ls -la` seems correct, though? Are you sure the EPERM isn't from something else (SELinux being a prominent cause)?

**Reply by abitrolly** [reply]
@mheon `--security-opt label=disable` should turn SELinux off, no?

Is that repeatable on your system? Mine is Fedora 36.

```sh
mkdir _delme
ls -laZ _delme
cd _delme
mkdir xxx
touch xxx/aaa
ls -laZ xxx
```
This outputs the following.
```
total 16
drwxr-xr-x.   2 anatoli anatoli unconfined_u:object_r:user_home_t:s0      4096 Oct 23 14:08 .
drwx--x--x. 125 anatoli anatoli unconfined_u:object_r:user_home_dir_t:s0 12288 Oct 23 14:08 ..
total 8
drwxr-xr-x. 2 anatoli anatoli unconfined_u:object_r:user_home_t:s0 4096 Oct 23 14:08 .
drwxr-xr-x. 3 anatoli anatoli unconfined_u:object_r:user_home_t:s0 4096 Oct 23 14:08 ..
-rw-r--r--. 1 anatoli anatoli unconfined_u:object_r:user_home_t:s0    0 Oct 23 14:08 aaa
```

Now running the container.
```sh
podman run -it --rm --security-opt label=disable --userns=keep-id -v $PWD/xxx:/xxx docker.io/gitpod/workspace-base
```
```sh
gitpod ~ $ ls -laZ /xxx
total 8
drwxr-xr-x.  2 anatoli anatoli unconfined_u:object_r:user_home_t:s0 4096 Oct 23 11:08 .
dr-xr-xr-x. 19 root    root    system_u:object_r:fusefs_t:s0        4096 Oct 23 11:43 ..
-rw-r--r--.  1 anatoli anatoli unconfined_u:object_r:user_home_t:s0    0 Oct 23 11:08 aaa
gitpod ~ $ touch /xxx/bbb
touch: cannot touch '/xxx/bbb': Permission denied
```

**Reply by mheon** [reply]
I can't grab the Gitpod image at the moment (Dockerhub unhelpfully giving me a 401), but using fedora:36 everything works as I expect.

**Reply by abitrolly** [reply]
I bet you need the image. Please try with from GitLab. It fails to `touch /xxx/111` too.
```
podman run -it --rm --security-opt label=disable --userns=keep-id -v $PWD/xxx:/xxx registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable
```

**Reply by rhatdan** [reply]
The process running within the container is running as gitpod, your user is not gitpod, so it is not allowed to write to the directory.

Use the :U option on the volume mount to fix the ownership of the directory.

**Reply by rhatdan** [reply]
```
$ podman run -it --rm --security-opt label=disable --userns=keep-id -v $PWD/xxx:/xxx:U registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable
gitpod ~ $ touch /xxx/dan
gitpod ~ $ exit
$ ls -ld xxx
drwxr-xr-x. 2 133332 133332 17 Oct 24 10:19 xxx
```

**Reply by abitrolly** [reply]
:U doesn't work, because it modifies file system permissions on the host filesystem, and as a result I lose write access to this dir from the host.

**Comment by eriksjolund** [comment]
@abitrolly  I think you need to use `--userns=keep-id:uid=$uid,gid=$gid` if you want to map the container user __gitpod__ to your regular user on the host. The new options `uid` and `gid` are only available in  the newly released [Podman 4.3](https://github.com/containers/podman/releases/tag/v4.3.0).
The container user __gitpod__ has the UID 33333 and the GID 33333, so you would need to specify   `--userns=keep-id:uid=33333,gid=33333`. You don't need to use the `:U` option.

If your Podman version is older than Podman 4.3, you could achieve the same thing by using __--uidmap__ and __--gidmap__ (see tip _Podman run fails with "Error: unrecognized namespace mode keep-id:uid=1000,gid=1000 passed"_ in [troubleshooting.md](https://github.com/containers/podman/blob/main/troubleshooting.md#39-podman-run-fails-with-error-unrecognized-namespace-mode-keep-iduid1000gid1000-passed))


I tested it out on Fedora 36 with Podman 4.2.1

1. Run `sudo useradd test`
2. Run `sudo machinectl shell test@`

3. Run these commands
```
[test@linux ~]$ podman pull -q registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable 
[test@linux ~]$ podman image inspect --format '{{.Config.User}} ' registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable
gitpod 
[test@linux ~]$ podman run --rm registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable /bin/grep gitpod /etc/passwd
gitpod:x:33333:33333::/home/gitpod:/bin/bash
```

4. Run `podman system reset`   (I don't know if this step is needed. See the discussion below)

5. Run these commands
```
[test@linux ~]$ uid=33333
[test@linux ~]$ gid=33333
[test@linux ~]$ subuidSize=$(( $(podman info --format "{{ range .Host.IDMappings.UIDMap }}+{{.Size }}{{end }}" ) - 1 ))
[test@linux ~]$ echo $subuidSize 
65536
[test@linux ~]$ subgidSize=$(( $(podman info --format "{{ range .Host.IDMappings.GIDMap }}+{{.Size }}{{end }}" ) - 1 ))
[test@linux ~]$ echo $subgidSize
65536
[test@linux ~]$ mkdir xxx
[test@linux ~]$ ls -ld xxx
drwxr-xr-x. 2 test test 18 Oct 25 18:53 xxx
[test@linux ~]$ podman run \
    --user $uid:$gid \
    --uidmap $uid:0:1 \
    --uidmap 0:1:$uid \
    --uidmap $(($uid+1)):$(($uid+1)):$(($subuidSize-$uid)) \
    --gidmap $gid:0:1 \
    --gidmap 0:1:$gid \
    --gidmap $(($gid+1)):$(($gid+1)):$(($subgidSize-$gid)) \
    -it \
    --rm \
    -v $PWD/xxx:/xxx:Z \
        registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable
gitpod ~ $ ls -ld /xxx
drwxr-xr-x. 2 gitpod gitpod 6 Oct 25 16:53 /xxx
gitpod ~ $ touch /xxx/file
gitpod ~ $ exit
exit
[test@linux ~]$ ls -ld xxx
drwxr-xr-x. 2 test test 18 Oct 25 18:54 xxx
[test@linux ~]$ ls -ld xxx/file
-rw-r--r--. 1 test test 0 Oct 25 18:54 xxx/file
[test@linux ~]$
```

First I tried the same thing without doing step 4 (i.e. running `podman system reset`) but then the `podman run ...` command in step 5 took forever. (I pressed Ctrl-C after a few minutes because I didn't want to wait for the command to finish).

This is the error message I got after pressing Ctrl-C

```
^CError: error creating container storage: creating an ID-mapped copy of layer "b954d3d642f12b6d10ac13b01a1985e03cbdce66dea86df312823dcd88e65b92": signal: interrupt
```

**Reply by abitrolly** [reply]
@eriksjolund thanks for the possible solution. Is it possible to add this as `--userns=map-id` to `podman` itself? https://docs.podman.io/en/latest/markdown/podman-run.1.html#userns-mode There is no way I could type it by hand every time.

**Reply by eriksjolund** [reply]
@abitrolly  I guess you are proposing that a new option `--userns=map-id` could be implemented that would instruct Podman to extract the UID and GID from the `.Config.User`  somehow

```
[test@linux ~]$ podman image inspect --format '{{.Config.User}} ' registry.gitlab.com/gitlab-org/gitlab-development-kit/gitpod-workspace:stable
gitpod 
```
and then provide the information as `--userns keep-id:uid=$uid,gid=$gid`.
It would be less to type and you would not need to look up the UID and GID.

**Reply by abitrolly** [reply]
@eriksjolund exactly this! Thanks for clarifying.

I would also limit this mapping only for filesystem access, but I am not sure Linux namespaces work this way.

**Reply by abitrolly** [reply]
@eriksjolund the option `--userns keep-id:uid=$uid` doesn't work with user name. How to get UID out of it?

```
$ podman run -it --rm --security-opt label=disable --userns=keep-id:uid=gitpod --name gitpod docker.io/gitpod/workspace-base     
Error: strconv.ParseUint: parsing "gitpod": invalid syntax
```

**Reply by eriksjolund** [reply]
```
$ podman run --rm -ti --name gitpod docker.io/gitpod/workspace-base /bin/grep gitpod /etc/passwd
WARNING: image platform ({amd64 linux  [] }) does not match the expected platform ({arm64 linux  [] })
gitpod:x:33333:33333::/home/gitpod:/bin/bash
$ 
```
UID is 33333 and GID is 33333

**Reply by abitrolly** [reply]
That two step execution doesn't seem reliable for all Linux flavors. Is it possible to extract that `UID` in `Dockerfile`? It allows `USER` to be numeric, and the spec says it could be used automatically then https://github.com/opencontainers/image-spec/blob/fe0a24978a6629f4b7159928e538dda36c7cec8e/conversion.md#configuser.

**Reply by abitrolly** [reply]
@eriksjolund is there a way to read `/etc/passwd` in from the image without running container?

**Reply by eriksjolund** [reply]
@abitrolly One way is described in https://unix.stackexchange.com/a/370221/9360
Somewhat modified it becomes:

```
container_id=$(podman create "$image")
podman cp "$container_id:/etc/passwd" "$destination_path"
podman rm "$container_id"
```

**Reply by abitrolly** [reply]
@eriksjolund I am going to try it. Thanks.

Is it possible to record the numeric UID into `Dockerfile` USER directive at the build time?

**Reply by abitrolly** [reply]
@eriksjolund seems to work fine. https://github.com/yakshaveinc/linux/blob/master/bin/runin-podman.sh 

Although I have concerns about how `--userns=keep-id` works. Looks like it runs container with my UID? What security issues can arise from that? I can run `sudo` and access some services that default users can't.

**Reply by rhatdan** [reply]
In containers we map the users UID into the container by default.  It is either mapped in as UID==0  or as UID==$UID (the users UID, with keep-id)  If you do not want the users UID mapped at all, you can use nomap.

```
$ id -u
3267
$ podman run --rm fedora cat /proc/self/uid_map 
         0       3267          1
         1     100000      65536
$ podman run --userns=keep-id --rm fedora cat /proc/self/uid_map 
         0          1       3267
      3267          0          1
      3268       3268      62269
$ podman run --userns=nomap --rm fedora cat /proc/self/uid_map 
         0          1      65536
```

The mappings get a little confusing here, since they are remapping the default user namespace.  A better way to show what is going on would be here:
```
$ podman run -d --rm fedora sleep 100
8988c2979ab3548753d7b2947a0261c9eaa31a7c651b0ba5a8411822c9617df3
kube (volume) $ podman top -l user huser 
USER        HUSER
root        3267
$ podman run -d --userns=keep-id --rm fedora sleep 100
08cf48b71a137974ca07a690a39f5e308849eb7ae12c2a16849f638939e4d22d
$ podman top -l user huser 
USER        HUSER
dwalsh      dwalsh
$ podman run -d --userns=nomap --rm fedora sleep 100
cc09a876879dc4604ca2656da303ff06e54088fd62794a4277d7b170586ca461
$ podman top -l user huser 
USER        HUSER
root        100000
```

**Reply by rhatdan** [reply]
Finally another example of the three mappings. I create a directory with a file in my home directory and then mount it into the container.  If I look at the file in the three different modes, the file ownership changes depending on the mapping.

```
$ mkdir test
$ touch test/foobar
$ podman run -v ./test:/test:Z --rm fedora ls -l /test
total 0
-rw-r--r--. 1 root root 0 Dec 29 12:20 foobar
$ podman run --userns=keep-id -v ./test:/test:Z --rm fedora ls -l /test
total 0
-rw-r--r--. 1 dwalsh dwalsh 0 Dec 29 12:20 foobar
$ podman run --userns=nomap -v ./test:/test:Z --rm fedora ls -l /test
total 0
-rw-r--r--. 1 nobody nobody 0 Dec 29 12:20 foobar
```

**Reply by abitrolly** [reply]
@rhatdan hi. Have you made any progress with the article about UID mappings? I am particularly interested in security implications and attack vectors depending on various combinations of these.

**Comment by rhatdan** [comment]
BTW I will be converting this into a Blog.

**Reply by abitrolly** [reply]
Without diagrams all UID mapping articles are mostly useless to me. When authors write something like `UID=$UID`, it all makes sense to them, because they already know which UID is where, but for users who don't even understand what there is some UID inside and UID outside, it is often doesn't make any sense. Curse of knowledge.

Like when you say "map", do you really mean that there is some shared memory with UID that is mapped inside of container? Or it is just associative array somewhere in Linux kernel, which is called "map" in some languages (dict in Python) which is a lookup table to find out which host UID to use to run container process. There are too many questions and that's makes it hard.

As a good example I would say that this article helped greatly https://www.redhat.com/en/blog/understanding-root-inside-and-outside-container It is not a diagram, but a good comparison table.

**Reply by eriksjolund** [reply]
>  Or it is just associative array 



I found some explanation in
the article _Minimizing Privilege for Building HPC Containers_ (Reid Priedhorsky, R. Shane Canon, Timothy Randles, and Andrew J. Younge)
https://arxiv.org/pdf/2104.07508.pdf 


_2.1.1 UID and GID maps._ 

quote:
_Because the map is one-to-one, there is no squashing of multiple IDs to a single ID in either direction._

quote:
_These mappings need not be complete. The four possibilities for a given pair of IDs are_
